gtsfm/view_graph_estimator/cycle_consistent_rotation_estimator.py

```diff
@@ -0,0 +1,231 @@
+
+import gtsfm.utils.geometry_comparisons as comp_utils
+import gtsfm.utils.graph as graph_utils
+import gtsfm.utils.logger as logger_utils
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.two_view_estimator import TwoViewEstimationReport
+from gtsfm.view_graph_estimator.view_graph_estimator_base import ViewGraphEstimatorBase
+
+logger = logger_utils.get_logger()
+
+ERROR_THRESHOLD = 7.0
+
+MAX_INLIER_MEASUREMENT_ERROR_DEG = 5.0
+
+
+class EdgeErrorAggregationCriterion(str, Enum):
+    """Aggregate cycle errors over each edge by choosing one of the following summary statistics:
+
+    MIN: Choose the mininum cycle error of all cyles this edge appears in. An edge that appears in ANY cycle
+        with low error is accepted. High recall, but can have low precision, as false positives can enter
+        (error was randomly cancelled out by another error, meaning accepted).
+    MEDIAN: Choose the median cycle error. robust summary statistic. At least half of the time, this edge
+        appears in a good cycle (i.e. of low error). Note: preferred over mean, which is not robust to outliers.
+
+    Note: all summary statistics will be compared with an allowed upper bound/threshold. If they exceed the
+    upper bound, they will be rejected.
+    """
+
+    MIN_EDGE_ERROR = "MIN_EDGE_ERROR"
+    MEDIAN_EDGE_ERROR = "MEDIAN_EDGE_ERROR"
+
+
+class CycleConsistentRotationViewGraphEstimator(ViewGraphEstimatorBase):
+    """A ViewGraphEstimator that filters two-view edges with high rotation-cycle consistency error.
+
+    The rotation cycle consistency error is computed by composing two-view relative rotations along a triplet, i.e:
+        inv(i2Ri0) * i2Ri1 * i1Ri0, where i2Ri0, i2Ri1 and i1Ri0 are 3 two-view relative rotations.
+
+    For each edge, the cyclic rotation error is computed for all 3-edge cycles (triplets) that it is a part of, which
+    are later aggregated using an EdgeErrorAggregationCriterion into a single value. The edge is added into the
+    ViewGraph is the aggregated error is less than a threshold.
+
+    Based off of:
+        https://github.com/sweeneychris/TheiaSfM/blob/master/src/theia/sfm/filter_view_graph_cycles_by_rotation.cc
+
+    See also:
+        C. Zach, M. Klopschitz, and M. Pollefeys. Disambiguating visual relations using loop constraints. In CVPR, 2010
+        http://people.inf.ethz.ch/pomarc/pubs/ZachCVPR10.pdf
+
+        Enqvist, Olof; Kahl, Fredrik; Olsson, Carl. Non-Sequential Structure from Motion. ICCVW, 2011.
+        https://portal.research.lu.se/ws/files/6239297/2255278.pdf
+    """
+
+    def __init__(
+        self,
+        edge_error_aggregation_criterion: EdgeErrorAggregationCriterion,
+        error_threshold: float = ERROR_THRESHOLD,
+    ) -> None:
+        self._edge_error_aggregation_criterion = edge_error_aggregation_criterion
+        self._error_threshold = error_threshold
+
+    def run(
+        self,
+        i2Ri1: Dict[Tuple[int, int], Rot3],
```

Comment:
@akshay-krishnan I think we should name this arg `i2Ri1_dict`, since `i2Ri1`means a single rotation elsewhere, but here we mean a dict